File: sonar-plugin-api/src/main/java/org/sonar/api/server/authentication/IdentityProvider.java
/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.api.server.authentication;

import org.sonar.api.ExtensionPoint;
import org.sonar.api.server.ServerSide;

/**
 * Entry-point to define a new Identity provider.
 * Only one of this two interfaces can be used :
 * <ul>
 *   <li>{@link OAuth2IdentityProvider} for OAuth2 authentication</li>
 *   <li>{@link BaseIdentityProvider} for other kind of authentication</li>
 * </ul>
 *
 * @since 5.4
 */
@ServerSide
@ExtensionPoint
public interface IdentityProvider {

  /**
   * Unique key of provider, for example "github".
   * Must not be blank.
   */
  String getKey();

  /**
   * Name displayed in login form.
   * Must not be blank.
   */
  String getName();

  /**
   * Display information for the login form
   */
  Display getDisplay();

  /**
   * Is the provider fully configured and enabled ? If {@code true}, then
   * the provider is available in login form.
   */
  boolean isEnabled();

  /**
   * Can users sign-up (connecting with their account for the first time) ? If {@code true},
   * then users can register and create their account into SonarQube, else only already
   * registered users can login.
   */
  boolean allowsUsersToSignUp();

}

File: server/sonar-web/src/main/js/apps/settings/components/authentication/queries/IdentityProvider.ts
/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { useContext } from 'react';
import {
  activateGithubProvisioning,
  activateScim,
  deactivateGithubProvisioning,
  deactivateScim,
  fetchIsScimEnabled,
} from '../../../../../api/provisioning';
import { getSystemInfo } from '../../../../../api/system';
import { AvailableFeaturesContext } from '../../../../../app/components/available-features/AvailableFeaturesContext';
import { useSyncStatusQuery } from '../../../../../queries/github-sync';
import { Feature } from '../../../../../types/features';
import { SysInfoCluster } from '../../../../../types/types';

export function useIdentityProvierQuery() {
  return useQuery(['identity_provider'], async () => {
    const info = (await getSystemInfo()) as SysInfoCluster;
    return { provider: info.System['External Users and Groups Provisioning'] };
  });
}

export function useScimStatusQuery() {
  const hasScim = useContext(AvailableFeaturesContext).includes(Feature.Scim);

  return useQuery(['identity_provider', 'scim_status'], () => {
    if (!hasScim) {
      return false;
    }
    return fetchIsScimEnabled();
  });
}

export function useGithubStatusQuery() {
  const hasGithubProvisioning = useContext(AvailableFeaturesContext).includes(
    Feature.GithubProvisioning
  );

  const res = useSyncStatusQuery({ enabled: hasGithubProvisioning });

  return { ...res, data: res.data?.enabled };
}

export function useToggleScimMutation() {
  const client = useQueryClient();
  return useMutation({
    mutationFn: (activate: boolean) => (activate ? activateScim() : deactivateScim()),
    onSuccess: () => {
      client.invalidateQueries({ queryKey: ['identity_provider'] });
    },
  });
}

export function useToggleGithubProvisioningMutation() {
  const client = useQueryClient();
  return useMutation({
    mutationFn: (activate: boolean) =>
      activate ? activateGithubProvisioning() : deactivateGithubProvisioning(),
    onSuccess: () => {
      client.invalidateQueries({ queryKey: ['identity_provider'] });
      client.invalidateQueries({ queryKey: ['github_sync'] });
    },
  });
}
